### PR TITLE
New spread-filter tool

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This tool is used to get the list of tests to execute following the rules defined for the project and based on a set of rules

This change includes a test and a config file which is similar to the final configuration that will be used in snapd project

### CODE REVIEW

File: utils/spread-filter [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            all_params.append(self.backend + ":" + self.system + ":" + exec_path)
+        return all_params
+
+    def _calc_self(self, change_path: str) -> str:
+        if self.verbose:
+            print("calculating self for {}".format(change_path))
+
+        # if the change is a directory which has changed
+        if os.path.isdir(change_path):
+            # if
```
Comment (zyga) [inline utils/spread-filter]: Truncated comment?

File: utils/spread-filter [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            if regex.match(path):
+                return self.to
+
+        return []
+
+
+class ExecutionRules:
+    def __init__(self, rules_map: dict[str, dict[str, dict[str, list[str]]]]):
+        self.rules = []
+        rules = list(rules_map["rules"].values())
```
Comment (zyga) [inline utils/spread-filter]: This is a weird API, code like this should be done on the side of the caller or in a helper like I suggested below.

```py
class ExecutionRules:
  def __init__(self, rules: list[ExecutionRule]):
    self.rules = rules

  @classmethod
  def from_rules_map(cls, rules_map: dict[str, dict[str, dict[str, list[str]]]]) -> Self:
      return cls([ExecutionRule(r.get("from", []), r.get("to", [])) for r in rules_map["rules"].values()])
```

**Review by zyga — COMMENTED** [review]
Looks good. One remark about a missing comment and one remark about API design.

Having a few unit tests would really help to see what's going on though. Integration tests are always slower to run.

File: utils/spread-filter (hunk 2 above)
Comment (sergiocazzolato) [reply utils/spread-filter]: updated

**Pull request merged**